In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# 이거 실행하고 다시 런타임 시작
# 한글 폰트
!sudo apt-get install -y fonts-nanum
!sudo fc-cache -fv
!rm ~/.cache/matplotlib -rf

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'sudo apt autoremove' to remove it.
The following NEW packages will be installed:
  fonts-nanum
0 upgraded, 1 newly installed, 0 to remove and 27 not upgraded.
Need to get 9,604 kB of archives.
After this operation, 29.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 fonts-nanum all 20170925-1 [9,604 kB]
Fetched 9,604 kB in 1s (6,912 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletyp

In [3]:
!pip install openpyxl
!pip install pykrx
!pip install opendartreader
!pip install --upgrade opendartreader

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 125 kB 8.7 MB/s 
     |████████████████████████████████| 52 kB 973 kB/s 
     |████████████████████████████████| 254 kB 44.8 MB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
import pandas as pd
import numpy as np
from pykrx import stock
import re
pd.options.display.max_columns = None

# KisValue 데이터 가공
- KisValue 데이터에서 2018~2021 코스닥 재무제표 가져오기

In [5]:
df = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-2021_원본.csv')

In [6]:
df.head(10)

Massive Download Unnamed: 1 Unnamed: 2     Unnamed: 3      Unnamed: 4  \
0                 NaN        NaN        NaN            NaN             NaN   
1  Portfolio - kosdaq        NaN        NaN            NaN             NaN   
2                 NaN        NaN        NaN            NaN             NaN   
3                 NaN        NaN        NaN            NaN             NaN   
4                 NaN        NaN        NaN    2018/Annual     2018/Annual   
5                 KIS      Stock       Name  총자산증가율/091010  자기자본증가율/091050   
6              036644     389500  (주)에스비비테크          46.03          -39.46   
7              PM6283     441270    (주)파인엠텍            NaN             NaN   
8              497730     296640    (주)이노룰스          81.84          215.02   
9              499459     417970   모델솔루션(주)          23.53           44.46   

      Unnamed: 5      Unnamed: 6             Unnamed: 7     Unnamed: 8  \
0            NaN             NaN                    NaN            NaN   
1            NaN             NaN                    NaN            NaN   
2            NaN             NaN                    NaN            NaN   
3            NaN             NaN                    NaN            NaN   
4    2018/Annual     2018/Annual            2018/Annual    2018/Annual   
5  매출액증가율/091060  영업이익증가율/091065  법인세비용차감전순이익증가율/091070  순이익증가율/091080   
6           2.26             NaN                    NaN            NaN   
7            NaN             NaN                    NaN            NaN   
8          -6.23          -67.95                 -81.56         -76.96   
9           9.65           73.99                 135.39         136.48   

       Unnamed: 9      Unnamed: 10             Unnamed: 11     Unnamed: 12  \
0             NaN              NaN                     NaN             NaN   
1             NaN              NaN                     NaN             NaN   
2             NaN              NaN                     NaN             NaN   
3             NaN              NaN                     NaN             NaN   
4     2018/Annual      2018/Annual             2018/Annual     2018/Annual   
5  종업원수증가율/091090  총자본영업이익율/092005  총자본법인세비용차감전순이익율/092010  총자본순이익율/092020   
6           16.88            -5.23                   -9.21           -9.21   
7             NaN              NaN                     NaN             NaN   
8             NaN             3.88                    1.19            1.43   
9            4.30            25.30                   24.73           19.34   

               Unnamed: 13      Unnamed: 14             Unnamed: 15  \
0                      NaN              NaN                     NaN   
1                      NaN              NaN                     NaN   
2                      NaN              NaN                     NaN   
3                      NaN              NaN                     NaN   
4              2018/Annual      2018/Annual             2018/Annual   
5  자기자본법인세비용차감전순이익율/092060  자기자본순이익율/092070  자본금법인세비용차감전순이익율/092080   
6                   -28.93           -28.92                 -168.79   
7                      NaN              NaN                     NaN   
8                     3.43             4.11                   57.29   
9                    55.15            43.14                  982.56   

      Unnamed: 16             Unnamed: 17     Unnamed: 18      Unnamed: 19  \
0             NaN                     NaN             NaN              NaN   
1             NaN                     NaN             NaN              NaN   
2             NaN                     NaN             NaN              NaN   
3             NaN                     NaN             NaN              NaN   
4     2018/Annual             2018/Annual     2018/Annual      2018/Annual   
5  자본금순이익율/092090  매출액법인세비용차감전순이익율/092100  매출액순이익율/092110  매출액영업이익율/092120   
6         -168.76                  -20.80          -20.80           -11.80   
7             NaN                     NaN             NaN         

In [ ]:
#     # 총자본회전율
#     tmp['총자본회전율'] = (tmp['매출액'] / tmp['자산총계']) * 100

#     # 자기자본회전율
#     tmp['자기자본회전율'] = (tmp['매출액'] / tmp['자본총계']) * 100

## 위에 두개는 수치가 안맞는다.

In [7]:
df

Massive Download Unnamed: 1 Unnamed: 2   Unnamed: 3   Unnamed: 4  \
0                    NaN        NaN        NaN          NaN          NaN   
1     Portfolio - kosdaq        NaN        NaN          NaN          NaN   
2                    NaN        NaN        NaN          NaN          NaN   
3                    NaN        NaN        NaN          NaN          NaN   
4                    NaN        NaN        NaN  2018/Annual  2018/Annual   
...                  ...        ...        ...          ...          ...   
1709                 NaN        NaN        NaN          NaN          NaN   
1710                 NaN        NaN        NaN          NaN          NaN   
1711                 NaN        NaN        NaN          NaN          NaN   
1712                 NaN        NaN        NaN          NaN          NaN   
1713                 NaN        NaN        NaN          NaN          NaN   

       Unnamed: 5   Unnamed: 6   Unnamed: 7   Unnamed: 8   Unnamed: 9  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4     2018/Annual  2018/Annual  2018/Annual  2018/Annual  2018/Annual   
...           ...          ...          ...          ...          ...   
1709          NaN          NaN          NaN          NaN          NaN   
1710          NaN          NaN          NaN          NaN          NaN   
1711          NaN          NaN          NaN          NaN          NaN   
1712          NaN          NaN          NaN          NaN          NaN   
1713          NaN          NaN          NaN          NaN          NaN   

      Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  Unnamed: 14  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4     2018/Annual  2018/Annual  2018/Annual  2018/Annual  2018/Annual   
...           ...          ...          ...          ...          ...   
1709          NaN          NaN          NaN          NaN          NaN   
1710          NaN          NaN          NaN          NaN          NaN   
1711          NaN          NaN          NaN          NaN          NaN   
1712          NaN          NaN          NaN          NaN          NaN   
1713          NaN          NaN          NaN          NaN          NaN   

      Unnamed: 15  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4     2018/Annual  2018/Annual  2018/Annual  2018/Annual  2018/Annual   
...           ...          ...          ...          ...          ...   
1709          NaN          NaN          NaN          NaN          NaN   
1710          NaN          NaN          NaN          NaN          NaN   
1711          NaN          NaN          NaN          NaN          NaN   
1712          NaN          NaN          NaN          NaN          NaN   
1713          NaN          NaN          NaN          NaN          NaN   

      Unnamed: 20  Unnamed: 21  Unnamed: 22  Unnamed: 23  Unnamed: 24  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4     2018/Annual  2018/Annual  2018/Annual  2018/Annual  2018/Annual   
...           ...         

In [9]:
# 재무제표 가공
df = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2018-2021_원본.csv')


def make_preprocessing_fs(df):

  # col_name = ['KIS', 'Stock', 'Name']
  # # kisvalue 데이터 다루기 편하게 handling
  # for i in df.columns[3:]:
  #   col_name.append(f"{df[i].iloc[5].split('/')[0]} ({df[i].iloc[4].split('/')[0]})")

  col_name = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    col_name.append(f"{df[i].iloc[5]} ({df[i].iloc[4].split('/')[0]})")

  df.columns = col_name
  df.drop([0,1,2,3,4,5], axis=0, inplace=True)
  df = df.reset_index(drop=True)


  # 뒤에 널값인 데이터도 같이 들어왔는데 같이 삭제
  last_index = df[df['Name']=='(주)아바텍'].index # 데이터가 들어있는 맨마지막 인덱스. 이 이후꺼를 삭제시키면 된다.
  df = df.iloc[:list(df[df['Name']=='(주)아바텍'].index).pop()+1, :]
  
  
  # 모든 재무제표가 비어있는 코드 drop
  df = df.T
  fs_all_null_code = [] # 모든 재무제표가 null값인 코드 저장을 위해
  for i in df.columns:
    if df[i].isna().sum()==len(df)-3: # KIS, Stock, Name을 제외한 모든 재무제표 컬럼의 개수
      fs_all_null_code.append(i)
  df.drop(fs_all_null_code, axis=1, inplace=True)
  df


  df = df.T # 다시 돌리기
  df = df.reset_index(drop = True)

  
  # stock code 없는 '에스케이제5호기업인수목적(주)' 삭제
  del_index = list(df[df['Stock'].isnull()].index)
  df.drop(del_index, axis=0,inplace=True)
  df= df.reset_index(drop= True)


  # 인수합병된 기업들 빼기
  df['인수합병여부'] = 0
  df['인수합병여부'] = df.apply(lambda x: 1 if '인수목적' in x['Name'] else 0, axis=1)


  df = df[df['인수합병여부']==0]
  df.drop('인수합병여부', axis=1, inplace=True)
  df = df.reset_index(drop = True)




  # 컬럼 selection ->삭제할거 삭제하기
  del_list = ['영업이익증가율', '비용차감전순이익증가율', '순이익증가율', '종업원수증가율', '총자본법인세비용차감전순이익율', '자기자본법인세비용차감전순이익율', '자기자본순이익율', '자본금법인세비용차감전순이익율', '매출액법인세비용차감전순이익율', '매출액(종업원1인당)', '법인세비용차감전순이익(종업원1인당)','순이익(종업원1인당)', '인건비(종업원1인당)' , '평균발행주식수', '유형,임대주택자산감가상각비','무형자산상각비', '매출액/121000', '무형자산(계)', '자본총계/118900', '[보통주발행주식수]', '[우선주배당금]' , '보통주배당금', '영업이익(손실)',  '유형고정자산증가율', '유동자산증가율', '재고자산증가율', '2021']

  drop_col_list = []
  for i in del_list:
    for j in df.columns:
      if i in j:
        drop_col_list.append(j)
  

  df.drop(drop_col_list,axis=1,inplace=True)

  # 컬럼명 다시 깔끔하게 만들기
  col_new = ['KIS', 'Stock', 'Name']
  for i in df.columns[3:]:
    start_index = i.find('/')
    col_new.append(i[:start_index] + i[start_index+7:])
  df.columns = col_new
  df.rename(columns={'유보액/093190 (2018)':'유보액/납입자본비율 (2018)', '유보액/093190 (2019)':'유보액/납입자본비율 (2019)', '유보액/093190 (2020)':'유보액/납입자본비율 (2020)'}, inplace = True)


  # 문자열 float로 변경
  for i in df.columns[3:]:
    df[i] = df[i].astype('str')
    df[i] = df[i].apply(lambda x: re.sub(",","",x))
    df[i] = df[i].astype('float')

  
  return df

  
df = make_preprocessing_fs(df)

In [10]:
(194.9 - 247.2)/247.2 * 100 

-21.156957928802584

In [11]:
df

KIS   Stock           Name  총자산증가율 (2018)  자기자본증가율 (2018)  \
0     036644  389500      (주)에스비비테크          46.03          -39.46   
1     497730  296640        (주)이노룰스          81.84          215.02   
2     499459  417970       모델솔루션(주)          23.53           44.46   
3     HV2035  393890       더블유씨피(주)          69.90           68.28   
4     HS8995  314140       (주)알피바이오          38.75           31.69   
...      ...     ...            ...            ...             ...   
1425  I73639  950110  SBI핀테크솔루션즈(주)         146.31           77.94   
1426  125028  046970         (주)우리로          42.14            8.83   
1427  383046  113810         디젠스(주)         -12.94            6.17   
1428  F11411  155650        와이엠씨(주)         -22.34           13.57   
1429  099145  149950         (주)아바텍          -2.81           -0.35   

      매출액증가율 (2018)  총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  \
0              2.26            -5.23           -9.21         -168.76   
1             -6.23             3.88            1.43           68.58   
2              9.65            25.30           19.34          768.48   
3               NaN            -8.72           -7.59         -559.62   
4             -5.02             8.28            5.66          146.00   
...             ...              ...             ...             ...   
1425          10.83             8.05            7.49           24.84   
1426          92.45             4.43           -0.55           -2.82   
1427         -29.81            -3.98          -16.02          -87.40   
1428          59.47            10.09            7.46          196.38   
1429         -22.27             3.76            5.22           94.76   

      매출액순이익율 (2018)  매출액영업이익율 (2018)  자기자본비율 (2018)  부채비율 (2018)  \
0             -20.80           -11.80          20.22       394.44   
1               1.90             5.16          40.95       144.17   
2              12.97            16.97          47.94       108.60   
3                NaN              NaN          93.30         7.18   
4               6.74             9.85          58.82        70.00   
...              ...              ...            ...          ...   
1425           16.69            17.92          45.68       118.91   
1426           -1.06             8.50          48.09       107.95   
1427          -31.83            -7.91          38.61       158.99   
1428            6.33             8.55          50.08        99.67   
1429           10.39             7.49          93.86         6.55   

      유보액/납입자본비율 (2018)  총자본회전율 (2018)  자기자본회전율 (2018)  당기순이익(손실) (2018)  \
0                340.07           0.44            1.39     -1.917031e+09   
1               2371.70           0.75            2.16      1.770280e+08   
2               2005.45           1.49            3.33      7.210181e+09   
3               8560.18            NaN             NaN     -9.513513e+09   
4               1558.22           0.84            1.40      4.363446e+09   
...                 ...            ...             ...               ...   
1425              93.99           0.45            0.88      3.963095e+09   
1426             194.50           0.52            0.96     -3.391840e+08   
1427              63.05           0.50            1.44     -1.185408e+10   
1428             800.95           1.18            2.87      1.464977e+10   
1429            1650.60           0.50            0.54      7.726745e+09   

      영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  자기자본증가율 (2019)  매출액증가율 (2019)  \
0            6.075800e+07         -21.16          -95.80         -34.70   
1            1.148444e+09          -1.20           32.63          43.95   
2            5.469434e+09           9.91           24.51           8.93   
3           -8.218473e+09         172.97           31.73            NaN   
4            6.995499e+09          29.38            7.64           7.86   
...                   ...            ...             ...            ...   
1425         2.682301e+09           

In [12]:
df.isna().sum()

KIS                     0
Stock                   0
Name                    0
총자산증가율 (2018)          45
자기자본증가율 (2018)         72
매출액증가율 (2018)          63
총자본영업이익율 (2018)        30
총자본순이익율 (2018)         29
자본금순이익율 (2018)         29
매출액순이익율 (2018)         46
매출액영업이익율 (2018)        46
자기자본비율 (2018)          48
부채비율 (2018)            48
유보액/납입자본비율 (2018)      29
총자본회전율 (2018)          55
자기자본회전율 (2018)         71
당기순이익(손실) (2018)       52
영업활동으로인한현금흐름 (2018)    62
총자산증가율 (2019)          31
자기자본증가율 (2019)         66
매출액증가율 (2019)          49
총자본영업이익율 (2019)        11
총자본순이익율 (2019)         12
자본금순이익율 (2019)         10
매출액순이익율 (2019)         31
매출액영업이익율 (2019)        30
자기자본비율 (2019)          30
부채비율 (2019)            30
유보액/납입자본비율 (2019)       9
총자본회전율 (2019)          37
자기자본회전율 (2019)         65
당기순이익(손실) (2019)       33
영업활동으로인한현금흐름 (2019)    34
총자산증가율 (2020)          12
자기자본증가율 (2020)         43
매출액증가율 (2020)          32
총자본영업이익율 (2020)         0
총자본순이익율 (2020)          0
자본금순이익율 (202

# 관리종목 연도별 종목코드 뽑기

In [13]:
# 2019, 2020, 2021, 2022
df_2019_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-1_수정.csv')
df_2019_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-2_수정.csv')
df_2019_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2019-3_수정.csv')

df_2020_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-1_수정.csv')
df_2020_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-2_수정.csv')
df_2020_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2020-3_수정.csv')

df_2021_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-1_수정.csv')
df_2021_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-2_수정.csv')
df_2021_3 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2021-3_수정.csv')

df_2022_1 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2022-1_수정.csv')
df_2022_2 = pd.read_csv('/content/drive/MyDrive/핀테크 파이널/data/2022-2_수정.csv')

In [14]:
# 관리대상기업 데이터 연도별 concat
df_2019 = pd.concat([df_2019_1, df_2019_2, df_2019_3], axis=0).reset_index(drop=True)
df_2020 = pd.concat([df_2020_1, df_2020_2, df_2020_3], axis=0).reset_index(drop=True)
df_2021 = pd.concat([df_2021_1, df_2021_2, df_2021_3], axis=0).reset_index(drop=True)
df_2022 = pd.concat([df_2022_1, df_2022_2], axis=0).reset_index(drop=True)

In [15]:
# 관리종목 지정된 종목코드 리스트 반환하는 함수
def return_risk_code(df):
  df['종목코드'] = df['종목코드'].astype('str')
  df['종목코드'] = df['종목코드'].apply(lambda x: x.zfill(6) if len(x)!=6 else x) # 종목코드 6자리로 만들어주기

  df['지정/해제여부'] = df.apply(lambda x: 0 if '해제' in x['공시제목'] else 1, axis=1)
  df = df[df['지정/해제여부']==1].reset_index(drop=True) # 관리종목 지정인 데이터만 불러오기
  
  종목코드_list = list(df['종목코드'].unique())
  return 종목코드_list



risk_code_2019 = return_risk_code(df_2019)
risk_code_2020 = return_risk_code(df_2020)
risk_code_2021 = return_risk_code(df_2021)
risk_code_2022 = return_risk_code(df_2022)

print(f'2019년도 관리종목기업코드 개수 : {len(risk_code_2019)}')
print(f'2020년도 관리종목기업코드 개수 : {len(risk_code_2020)}')
print(f'2021년도 관리종목기업코드 개수 : {len(risk_code_2021)}')
print(f'2022년도 관리종목기업코드 개수 : {len(risk_code_2022)}')

2019년도 관리종목기업코드 개수 : 93
2020년도 관리종목기업코드 개수 : 101
2021년도 관리종목기업코드 개수 : 98
2022년도 관리종목기업코드 개수 : 73


# 기존 KisValue데이터에 관리종목여부 추가
- 2019, 2020, 2021 여부만 추가할 것.
- 모델링할 때 데이터 불균형 문제를 어느정도 해결해야할 것 같다.

In [16]:
# 일단 모든 값 0으로 초기화
# 인수합병된 기업들을 미리 지워서 그런지 위에서 출력한 개수보다 적게 나온다.

df['2019_관리종목여부'] = 0
df['2020_관리종목여부'] = 0
df['2021_관리종목여부'] = 0


df['2019_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2019 else 0 , axis=1)
df['2020_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2020 else 0 , axis=1)
df['2021_관리종목여부'] = df.apply(lambda x: 1 if x['Stock'] in risk_code_2021 else 0 , axis=1)

print(df['2019_관리종목여부'].value_counts())
print(df['2020_관리종목여부'].value_counts())
print(df['2021_관리종목여부'].value_counts())

0    1388
1      42
Name: 2019_관리종목여부, dtype: int64
0    1392
1      38
Name: 2020_관리종목여부, dtype: int64
0    1400
1      30
Name: 2021_관리종목여부, dtype: int64


In [ ]:
# 위에꺼 확인용 코드
# tmp=[]
# for i in risk_code_2021:
#   if i in list(df['Stock'].values):
#     tmp.append(i)
# print(len(tmp))

In [ ]:
df.describe()

총자산증가율 (2018)  자기자본증가율 (2018)  매출액증가율 (2018)  총자본영업이익율 (2018)  \
count   1.385000e+03     1358.000000    1367.000000      1400.000000   
mean    8.307525e+02      685.906127      21.573921         2.502879   
std     2.926893e+04    23315.797840     153.383154        15.853033   
min    -6.804000e+01      -90.030000    -100.000000      -129.530000   
25%    -1.530000e+00       -1.100000      -8.145000        -0.775000   
50%     7.340000e+00        6.950000       4.380000         2.905000   
75%     2.496000e+01       23.480000      20.830000         8.742500   
max     1.089023e+06   858889.270000    3648.630000       137.560000   

       총자본순이익율 (2018)  자본금순이익율 (2018)  매출액순이익율 (2018)  매출액영업이익율 (2018)  \
count     1401.000000     1401.000000     1384.000000      1384.000000   
mean        -1.536881       70.205332     -616.313439      -471.265383   
std         26.836112     2323.153378    11080.499725     10273.463568   
min       -532.050000   -64738.350000  -353205.520000   -350051.850000   
25%         -3.400000      -31.970000       -4.920000        -0.890000   
50%          2.620000       33.020000        3.670000         4.455000   
75%          7.760000      143.810000       10.500000        11.710000   
max        114.400000    52574.620000     1307.340000        77.620000   

       자기자본비율 (2018)  부채비율 (2018)  유보액/납입자본비율 (2018)  총자본회전율 (2018)  \
count    1382.000000  1382.000000        1401.000000    1375.000000   
mean       64.496852    97.685753        1697.744404       0.802269   
std        20.955324   299.279136        9641.966500       0.633225   
min         1.120000     0.910000       -8327.170000       0.010000   
25%        49.792500    21.345000         334.930000       0.400000   
50%        65.125000    53.550000         783.650000       0.660000   
75%        82.407500   100.827500        1592.300000       1.050000   
max        99.100000  8799.540000      248880.480000       8.440000   

       자기자본회전율 (2018)  당기순이익(손실) (2018)  영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  \
count     1359.000000      1.378000e+03         1.368000e+03    1399.000000   
mean         1.610412      2.820362e+09         5.502024e+09      21.805661   
std          2.491764      1.698619e+10         1.958206e+10      77.475601   
min          0.010000     -1.234088e+11        -1.672801e+11     -82.480000   
25%          0.580000     -2.380625e+09        -1.445045e+09      -1.220000   
50%          1.050000      1.959088e+09         2.313819e+09       7.260000   
75%          1.880000      6.327230e+09         8.111196e+09      22.825000   
max         60.310000      1.546099e+11         2.871206e+11    1628.770000   

       자기자본증가율 (2019)  매출액증가율 (2019)  총자본영업이익율 (2019)  총자본순이익율 (2019)  \
count     1364.000000    1381.000000      1419.000000     1418.000000   
mean        32.059802     116.543526         1.872854       -2.697045   
std        173.069248    3535.374771        16.070914       25.882170   
min        -98.300000    -100.000000      -191.540000     -353.270000   
25%         -1.172500      -8.560000        -0.425000       -3.970000   
50%          5.500000       5.300000         2.810000        2.200000   
75%         19.890000      21.100000         8.230000        7.217500   
max       3408.850000  131338.880000        89.070000       71.040000   

       자본금순이익율 (2019)  매출액순이익율 (2019)  매출액영업이익율 (2019)  자기자본비율 (2019)  \
count     1420.000000     1399.000000      1400.000000    1400.000000   
mean      -437.646775     -369.679006      -110.732171      64.143086   
std      19593.758488     6471.426055      1343.674254      20.981897   
min    -734161.410000  -184682.910000    -32961.540000       1.080000   
25%        -35.380000       -7.115000        -0.497500      49.122500   
50%         27.930000        3.100000         4.490000      65.405000   
75%        126.837500       10.230000        11.542500      82.162500   
max      68510.460000     1085.710000        99.850000      99.570000   

       부채비율 (2019)  유보액

In [17]:
# 2019 관리종목대상의 2018 재무제표 평균 보기
print(df[df['2019_관리종목여부']==1]['총자산증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자본금순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['매출액영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['부채비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['유보액/납입자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['총자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['자기자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['당기순이익(손실) (2018)'].mean())
print(df[df['2019_관리종목여부']==1]['영업활동으로인한현금흐름 (2018)'].mean())

8.886904761904763
7.115714285714287
-3.3880952380952385
-8.333809523809524
-17.688333333333333
-106.68761904761905
-51.57119047619048
-26.022857142857145
49.95333333333333
142.62261904761905
564.6380952380953
0.5435714285714286
1.2719047619047619
-10126616951.219513
-1637915292.682927


In [18]:
# 위랑 비교해서 2019 관리종목대상의 2018 재무제표 평균 보기
print(df[df['2019_관리종목여부']==0]['총자산증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['매출액증가율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본영업이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자본금순이익율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['매출액순이익율 (2018)'].mean()) # 이거 아웃라이어 있는거같은데 살짝 지우자. 
print(df[df['2019_관리종목여부']==0]['매출액영업이익율 (2018)'].mean()) # 이것도 아웃라이어 데이터 삭제해주자.
print(df[df['2019_관리종목여부']==0]['자기자본비율 (2018)'].mean())

print(df[df['2019_관리종목여부']==0]['부채비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['유보액/납입자본비율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['총자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['자기자본회전율 (2018)'].mean())
print(df[df['2019_관리종목여부']==0]['영업활동으로인한현금흐름 (2018)'].mean())

856.4549218168281
707.5696504559271
22.365169811320758
2.838033873343152
-1.037718910963944
75.67222222222223
-633.9879359165425
-485.19994783904616
64.95269402985075
96.27728358208955
64.95269402985075
1732.7631420161883
0.8104201050262565
1.6212072892938498
5722624618.688771


2019년만 슬쩍 봤는데 평균으로 찍어봐도 유의미한 결과가 나온다.


## 일단 2018년 컬럼 모두가 null이면 drop(axis=0) 하자
- 2017년이나 2018년 이전부터 설립된 기업만을 대상으로 하기 위해
- 내가 확인한결과 2020, 2021은 모두 nan인게 없다.

In [19]:
df3 = df.iloc[:,3:18].T
all_nan_2018 = list(df3.isna().sum()[df3.isna().sum().values==15].index)

df = df.drop(all_nan_2018, axis=0).reset_index(drop=True)

In [20]:
print(df['2019_관리종목여부'].value_counts())
print(df['2020_관리종목여부'].value_counts())
print(df['2021_관리종목여부'].value_counts())

0    1359
1      42
Name: 2019_관리종목여부, dtype: int64
0    1363
1      38
Name: 2020_관리종목여부, dtype: int64
0    1372
1      29
Name: 2021_관리종목여부, dtype: int64


In [21]:
df.isna().sum()

KIS                     0
Stock                   0
Name                    0
총자산증가율 (2018)          16
자기자본증가율 (2018)         43
매출액증가율 (2018)          34
총자본영업이익율 (2018)         1
총자본순이익율 (2018)          0
자본금순이익율 (2018)          0
매출액순이익율 (2018)         17
매출액영업이익율 (2018)        17
자기자본비율 (2018)          19
부채비율 (2018)            19
유보액/납입자본비율 (2018)       0
총자본회전율 (2018)          26
자기자본회전율 (2018)         42
당기순이익(손실) (2018)       23
영업활동으로인한현금흐름 (2018)    33
총자산증가율 (2019)           2
자기자본증가율 (2019)         37
매출액증가율 (2019)          20
총자본영업이익율 (2019)         1
총자본순이익율 (2019)          2
자본금순이익율 (2019)          0
매출액순이익율 (2019)         13
매출액영업이익율 (2019)        12
자기자본비율 (2019)          21
부채비율 (2019)            21
유보액/납입자본비율 (2019)       0
총자본회전율 (2019)          19
자기자본회전율 (2019)         47
당기순이익(손실) (2019)       23
영업활동으로인한현금흐름 (2019)    24
총자산증가율 (2020)           3
자기자본증가율 (2020)         32
매출액증가율 (2020)          14
총자본영업이익율 (2020)         0
총자본순이익율 (2020)          0
자본금순이익율 (202

# 네이버 파이낸스로 채울수 있는부분 채워보기
- 우리는 연도를 기준으로 ??

<pre>
- 총자산증가율 : (올해 총자산 - 전년 총자산) / 전년 총자산 * 100
- 자기자본증가율 : (올해 자기자본 - 전년 자기자본)/전년 자기자본 * 100 
- 매출액증가율 : (올해 매출액 - 전년도 매출액) / 전년도 매출액 * 100 
- 총자본영업이익율 : (영업이익/총자산) * 100
- 총자본순이익율 : ROA랑 같다. (당기순이익/총자산) * 100
- 매출액순이익율 : ROS랑 같다. (당기순이익) / (매출액) * 100
- 매출액영업이익율 : 영업이익 / 매출액 * 100
- 자기자본비율 : BIS랑 같다. 자기자본 / 총자산 * 100
- 부채비율 : (부채총계 / 자본총계) * 100
- 총자본회전율 : (매출액 / 총자본 (== 총자산) ) * 100
- 자기자본회전율 : (매출액 / 자기자본 ) * 100
- 당기순이익(손실) : 이거 그냥 당기순이익으로 메꿔도 되나??
- 영업활동으로인한현금흐름 : CFO랑 같다. 영업이익 + 감가상각비 - 법인세


필요 재무제표 : 총자산, 자기자본, 매출액, 

In [22]:
df.head(30)

KIS   Stock          Name  총자산증가율 (2018)  자기자본증가율 (2018)  \
0   036644  389500     (주)에스비비테크          46.03          -39.46   
1   497730  296640       (주)이노룰스          81.84          215.02   
2   499459  417970      모델솔루션(주)          23.53           44.46   
3   HV2035  393890      더블유씨피(주)          69.90           68.28   
4   HS8995  314140      (주)알피바이오          38.75           31.69   
5   389238  368970       (주)오에스피            NaN             NaN   
6   IW9267  394280  오픈엣지테크놀로지(주)       24565.93        24767.99   
7   484917  148930    (주)에이치와이티씨           6.46             NaN   
8   894281  107600        (주)새빗켐          24.53           15.87   
9   867055  129920      (주)대성하이텍           8.36           26.30   
10  E66612  368600     (주)아이씨에이치          35.25          165.42   
11  IO9843  365340      성일하이텍(주)          42.94           47.71   
12  I91043  397030    (주)에이프릴바이오         137.78          -40.83   
13  J89214  328130         (주)루닛         154.08             NaN   
14  IO4904  403870    (주)에이치피에스피         -51.38          -43.67   
15  370082  112290      영창케미칼(주)           5.16            3.00   
16  166646  402030    (주)코난테크놀로지          18.78            2.69   
17  FZ6224  412350       레이저쎌(주)          25.26           42.92   
18  CR3994  310210       보로노이(주)         243.30          213.07   
19  CC9475  377460     (주)위니아에이드          63.52          820.48   
20  IW0365  362320      (주)청담글로벌         359.54          341.83   
21  J50025  399720       (주)가온칩스          65.42           73.30   
22  I66973  389260      대명에너지(주)         -30.60           12.01   
23  IT6914  389140       (주)포바이포         226.18           96.88   
24  G94198  388050       지투파워(주)          34.93           40.13   
25  219717  396300     (주)세아메카닉스          26.53           14.90   
26  F26242  366030       (주)공구우먼           8.65           26.41   
27  E82895  388720     (주)유일로보틱스          19.25            4.34   
28  L65171  288980      (주)모아데이타          40.62           83.50   
29  HL2329  376930         노을(주)          31.39           23.08   

    매출액증가율 (2018)  총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  \
0            2.26            -5.23           -9.21         -168.76   
1           -6.23             3.88            1.43           68.58   
2            9.65            25.30           19.34          768.48   
3             NaN            -8.72           -7.59         -559.62   
4           -5.02             8.28            5.66          146.00   
5             NaN            21.66           15.28          284.44   
6             NaN            26.41           25.40         2683.41   
7           48.83            21.61           14.00           74.47   
8            5.81            14.03           10.14           51.67   
9            7.67             3.96            1.57          162.88   
10          -5.52            20.84           19.83          864.26   
11          50.40            19.89           16.20          318.06   
12            NaN           -11.24           -9.94         -109.36   
13          71.88           -83.72         -102.02        -1967.71   
14         -91.62           -14.48          -12.67         -126.71   
15          40.27             7.28            1.55           92.82   
16          16.39             4.00            1.58            8.01   
17           2.47           -68.09          -67.44         -145.40   
18        -100.00           -39.23          -40.08         -173.61   
19         124.61            11.58            8.63          439.49   
20        3173.33            58.38           57.31          311.91   
21          72.07            17.28           18.32         1002.90   
22          -4.95             7.56            5.38           63.43   
23         506.28            43.40           41.30          739.42   
24          89.58            19.33           16.79          512.59   
25          44.22             9.27           10.63          684.60   
26           2.53            24.59     

# 네이버 파이낸스에 없는 데이터는 kisvalue에도 없는 것 같다.


# OpenDartAPI 사용해서 그나마 있는거 채우기
- opendartApi에는 연결재무제표 or 재무제표밖에 없다.

In [ ]:
# 일단 모두 연간으로 계산해야할 것 같다.

# 변화율을 구해야하므로 2017년부터 2020년 재무제표 데이터까지 구하고 비율 계산
# 총자산증가율

In [23]:
import OpenDartReader

### 0. 객체 생성 ###
# 객체 생성 (API KEY 지정) 
api_key = 'fed627c93ccbd6bbc2ad3c3769f4c41be9e377eb'

dart = OpenDartReader(api_key) 

result = dart.finstate('LG전자', 2020, '11011') # 2020년 사업보고서 내 재무 정보 가져오기
result = result.loc[result['fs_nm']=='연결재무제표'] # 연결재무제표 필터링

# account_nm이 당기순이익, 영업이익, 매출액 중 하나의 경우에만 가져오기
result = result.loc[result['account_nm'].isin(['당기순이익', '영업이익', '영업이익', '매출액'])]

# 열 필터링
result = result.loc[:,['account_nm', 'fs_nm', 'sj_nm', 'thstrm_dt', 'thstrm_amount']]
display(result)

account_nm   fs_nm  sj_nm                thstrm_dt       thstrm_amount
9         매출액  연결재무제표  손익계산서  2020.01.01 ~ 2020.12.31  63,262,046,000,000
10       영업이익  연결재무제표  손익계산서  2020.01.01 ~ 2020.12.31   3,194,987,000,000
12      당기순이익  연결재무제표  손익계산서  2020.01.01 ~ 2020.12.31   2,063,790,000,000

# 필요한 재무제표 데이터 불러오기
-  개별은 아래에서 재무제표만 뽑아내면 된다.
- indicator 다시 적용(추가로 구할수있는거 있으면 구하게)
- 11011(사업보고서) 만 갖고올수있게끔

In [24]:
# 기업 연도 한개씩으로 내가 새로 만들어보기
# 정민이한테 물어보기
stock_code = '018700'
year = 2017
indicators = ['자산총계', '부채총계', '자본총계', '매출액', '영업이익', '당기순이익', '법인세차감전 순이익']


result = dart.finstate(stock_code, year,'11011') # 사업보고서 데이터 가져오기
# if report is None: # 리포트가 없다면 (참고 : 리포트가 없으면 None을 반환함)
result = result.loc[result['fs_nm']=='재무제표'] # 연결재무제표 아닌 재무제표 필터링
result = result.loc[result['account_nm'].isin(indicators)]
data = [] 
for y, c in zip([year], ['thstrm_amount']):
  record = [stock_code, y]
  for indic in indicators:
    # account_nm이 indic인 행의 c 컬럼 값을 가져옴
    if sum(result['account_nm']==indic)>0: # 해당 지표가 있다면 추가(지표가 없는 경우가 있음)
      value = result.loc[result['account_nm']==indic, c].iloc[0]
    else:
      value = np.nan

    record.append(value)

  data.append(record)
tt= pd.DataFrame(data, columns = ['기업', '연도'] + indicators)  
tt

기업    연도            자산총계           부채총계            자본총계            매출액  \
0  018700  2017  40,475,076,097  2,607,763,705  37,867,312,392  3,499,939,883   

           영업이익         당기순이익    법인세차감전 순이익  
0  -604,655,894  -902,076,584  -112,999,374

In [25]:
# 함수화


def crawl_finance(stock_code, year, indicators):


  result = dart.finstate(stock_code, year,'11011') # 사업보고서 데이터 가져오기
  # if report is None: # 리포트가 없다면 (참고 : 리포트가 없으면 None을 반환함)
  result = result.loc[result['fs_nm']=='재무제표'] # 연결재무제표 아닌 재무제표 필터링
  result = result.loc[result['account_nm'].isin(indicators)]
  data = [] 
  for y, c in zip([year], ['thstrm_amount']):
    record = [stock_code, y]
    for indic in indicators:
      # account_nm이 indic인 행의 c 컬럼 값을 가져옴
      if sum(result['account_nm']==indic)>0: # 해당 지표가 있다면 추가(지표가 없는 경우가 있음)
        value = result.loc[result['account_nm']==indic, c].iloc[0]
      else:
        value = np.nan

      record.append(value)

    data.append(record)
  tt= pd.DataFrame(data, columns = ['기업', '연도'] + indicators)  
  tt
  return tt

In [26]:
year = [2017, 2018, 2019, 2020]
indicators = ['자산총계', '부채총계', '자본총계', '매출액', '영업이익', '당기순이익', '법인세차감전 순이익']

# 널값인 데이터만 보기
df2 = df.copy()
df2['NaN_cnt'] = df2.isnull().sum(1)

# 재무제표 crawling이 필요한 코드
stock_code = list(df2[df2['NaN_cnt']>0]['Stock'].values)
 
# 널값이 있는 종목코드 재무제표만 저장하는 데이터프레임
null_fill_data = pd.DataFrame(columns = ['기업', '연도'] + indicators) 
for i in stock_code:
  for j in year:
     try:
       tt = crawl_finance(i, j, indicators)
       null_fill_data = pd.concat([null_fill_data, tt], axis=0)
     except:
       pass

null_fill_data

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없습니다.'}

{'status': '013', 'message': '조회된 데이타가 없

기업    연도            자산총계            부채총계            자본총계  \
0   347700  2020   9,539,670,072   7,353,748,337   2,185,921,735   
0   338220  2020   9,587,094,279  13,229,235,096  -3,642,140,817   
0   334970  2020  50,232,970,362  51,630,407,000  -1,397,436,638   
0   340930  2020  40,790,146,981  20,485,252,169  20,304,894,812   
0   321820  2020   6,027,806,557   7,157,246,083  -1,129,439,526   
..     ...   ...             ...             ...             ...   
0   291230  2019  15,055,376,421     919,220,457  14,136,155,964   
0   291230  2020  15,152,725,341     945,135,877  14,207,589,464   
0   284620  2018  10,599,449,914     772,060,139   9,827,389,775   
0   284620  2019  10,697,358,214     789,886,882   9,907,471,332   
0   284620  2020  30,850,568,946  18,497,921,073  12,352,647,873   

               매출액             영업이익            당기순이익       법인세차감전 순이익  
0    2,665,429,178   -3,657,769,678   -9,308,741,731   -9,308,741,731  
0    1,257,218,734   -9,721,688,063  -11,108,598,035  -11,108,598,035  
0                -  -22,601,457,755  -31,033,281,007  -31,033,281,007  
0   58,978,131,009    7,718,043,746    4,765,270,114    5,240,984,389  
0   28,799,193,142   -3,232,982,943   -3,242,623,376   -3,242,623,376  
..             ...              ...              ...              ...  
0                -      -26,942,340      164,510,142      182,705,299  
0              NaN      -48,677,430       71,433,500       72,983,568  
0                0      -25,917,490       38,946,647       49,931,598  
0                0      -44,177,300       80,081,557       90,492,514  
0                -   -8,417,500,666  -13,340,652,897  -14,656,832,110  

[156 rows x 9 columns]

In [27]:
# 중복된 행 제거
null_fill_data.drop_duplicates(inplace = True)

In [28]:
# 숫자로 모두 변환
def str_to_float(value):
  if type(value)==float: # nan의 경우에는 문자가 아니라, 이미 float 취급됨
    return value
  elif value =='-': # -로 되어 있으면 0으로 변환
    return 0
  else:
    return float(value.replace(',',''))

for indc in indicators:
  null_fill_data[indc] = null_fill_data[indc].apply(str_to_float)

In [29]:
# 기업, 연도 기준으로 sorting
null_fill_data.sort_values(by=['기업', '연도'], inplace= True)
null_fill_data = null_fill_data.reset_index(drop=True)

In [30]:
null_fill_data.head(30)

기업    연도          자산총계          부채총계          자본총계           매출액  \
0   006730  2017  1.502578e+12  1.018991e+12  4.835862e+11  5.754273e+10   
1   006730  2018  1.724660e+12  1.076229e+12  6.484311e+11  1.081047e+11   
2   006730  2019  1.811212e+12  1.138809e+12  6.724037e+11  1.279366e+11   
3   006730  2020  1.818179e+12  1.170517e+12  6.476618e+11  8.410463e+10   
4   018700  2017  4.047508e+10  2.607764e+09  3.786731e+10  3.499940e+09   
5   018700  2018  3.850006e+10  2.637741e+09  3.586232e+10  8.732299e+09   
6   018700  2019  3.715530e+10  3.033724e+09  3.412158e+10  3.506172e+09   
7   018700  2020  5.118833e+10  1.034894e+10  4.083939e+10  2.464416e+10   
8   019570  2018  1.702666e+10  2.486696e+09  1.453996e+10           NaN   
9   019570  2019  2.594744e+10  5.790251e+09  2.015719e+10           NaN   
10  019570  2020  5.848363e+10  3.644073e+10  2.204290e+10           NaN   
11  020400  2017  6.938612e+10  4.177556e+10  2.761056e+10  7.470931e+10   
12  020400  2018  8.506983e+10  5.197385e+10  3.309598e+10  8.086266e+10   
13  020400  2019  1.065335e+11  7.192466e+10  3.460887e+10  8.852912e+10   
14  020400  2020  1.062067e+11  7.023129e+10  3.597539e+10  1.111199e+11   
15  021080  2017  7.865336e+10  7.701542e+09  7.095181e+10  2.512686e+10   
16  021080  2018  7.855023e+10  3.197076e+09  7.535316e+10  1.711689e+10   
17  021080  2019  8.411857e+10  4.417514e+09  7.970105e+10  2.144129e+10   
18  021080  2020  1.004025e+11  8.478039e+09  9.192447e+10  3.218911e+10   
19  023460  2017  1.357566e+11  3.445001e+10  1.013066e+11  6.465576e+09   
20  023460  2018  1.457583e+11  4.444101e+10  1.013173e+11  8.713904e+09   
21  023460  2019  1.516332e+11  4.320709e+10  1.084261e+11  2.191583e+10   
22  023460  2020  1.714962e+11  4.592012e+10  1.255760e+11  4.030024e+10   
23  035290  2017  1.699479e+10  3.890051e+09  1.310474e+10  1.280204e+10   
24  035290  2018  4.247263e+10  2.860562e+10  1.386701e+10  8.680452e+09   
25  035290  2019  1.609617e+10  2.109726e+10 -5.001092e+09  7.642733e+09   
26  035290  2020  2.063417e+10  6.145524e+09  1.448865e+10  4.138760e+09   
27  036170  2017  9.743241e+10  2.613492e+10  7.129749e+10  1.557025e+10   
28  036170  2018  1.137376e+11  3.567021e+10  7.806737e+10  1.419061e+10   
29  036170  2019  1.014129e+11  1.364875e+10  8.776414e+10  1.348926e+10   

            영업이익         당기순이익    법인세차감전 순이익  
0  -1.400903e+10 -2.655790e+10 -2.805165e+10  
1  -1.541956e+07 -1.037741e+10 -6.509073e+09  
2   1.443905e+10  4.709363e+07  2.180952e+10  
3  -6.472848e+09 -1.682075e+10 -1.777866e+10  
4  -6.046559e+08 -9.020766e+08 -1.129994e+08  
5  -2.019640e+09 -5.642155e+09 -5.642155e+09  
6  -3.320440e+09 -7.076383e+09 -7.076383e+09  
7   9.113775e+08 -8.545291e+09 -8.545291e+09  
8  -7.409675e+08 -1.060834e+09 -1.060834e+09  
9  -3.339529e+09 -3.261240e+09 -3.261240e+09  
10 -1.098521e+10 -1.111546e+10 -1.111546e+10  
11  1.607199e+09  8.369929e+08  1.068816e+09  
12  1.905642e+09  8.605013e+08  1.072223e+09  
13  3.262382e+09  7.618720e+05 -1.466504e+08  
14  2.941199e+09  1.523659e+09  1.722763e+09  
15  1.230264e+10  8.721283e+09  1.133059e+10  
16  8.021335e+09  7.019155e+09  7.879099e+09  
17  8.530282e+09  6.747625e+09  8.495202e+09  
18  1.719891e+10  1.462314e+10  1.702606e+10  
19  1.052597e+09  1.343563e+09  1.753727e+09  
20  2.175231e+09  1.405417e+09  2.179910e+09  
21  1.162201e+10  8.828807e+09  1.157796e+10  
22  3.246359e+10  2.663508e+10  3.258134e+10  
23 -3.690116e+09 -8.062623e+09 -8.023554e+09  
24 -4.559128e+09 -4.045846e+09 -4.714757e+09  
25 -7.142117e+09 -2.488321e+10 -2.489904e+10  
26 -2.343112e+09  1.635730e+09  1.396873e+09  
27  2.881709e+09  5.846181e+08 -1.803704e+08  
28  1.307041e+09 -2.572776e+09 -3.868633e+09  
29 -9.697300e+05 -1.528333e+10 -1.047195e+10

In [31]:
null_fill_data['자산총계'].diff(1)

0               NaN
1      2.220827e+11
2      8.655190e+10
3      6.966514e+09
4     -1.777704e+12
           ...     
151   -1.109239e+10
152   -3.125048e+10
153    4.613555e+10
154   -6.281865e+09
155   -4.934685e+10
Name: 자산총계, Length: 156, dtype: float64

In [32]:
null_fill_data['매출액'].shift(1)

0               NaN
1      5.754273e+10
2      1.081047e+11
3      1.279366e+11
4      8.410463e+10
           ...     
151    4.525714e+10
152    5.897813e+10
153    2.665429e+09
154    4.591485e+09
155    2.479277e+09
Name: 매출액, Length: 156, dtype: float64

In [33]:
null_fill_data

기업    연도          자산총계          부채총계          자본총계           매출액  \
0    006730  2017  1.502578e+12  1.018991e+12  4.835862e+11  5.754273e+10   
1    006730  2018  1.724660e+12  1.076229e+12  6.484311e+11  1.081047e+11   
2    006730  2019  1.811212e+12  1.138809e+12  6.724037e+11  1.279366e+11   
3    006730  2020  1.818179e+12  1.170517e+12  6.476618e+11  8.410463e+10   
4    018700  2017  4.047508e+10  2.607764e+09  3.786731e+10  3.499940e+09   
..      ...   ...           ...           ...           ...           ...   
151  340930  2020  4.079015e+10  2.048525e+10  2.030489e+10  5.897813e+10   
152  347700  2020  9.539670e+09  7.353748e+09  2.185922e+09  2.665429e+09   
153  348150  2020  5.567522e+10  1.541278e+09  5.413395e+10  4.591485e+09   
154  354200  2020  4.939336e+10  7.944209e+09  4.144915e+10  2.479277e+09   
155  950200  2020  4.651200e+07  1.228189e+07  3.423011e+07  1.610598e+07   

             영업이익         당기순이익    법인세차감전 순이익  
0   -1.400903e+10 -2.655790e+10 -2.805165e+10  
1   -1.541956e+07 -1.037741e+10 -6.509073e+09  
2    1.443905e+10  4.709363e+07  2.180952e+10  
3   -6.472848e+09 -1.682075e+10 -1.777866e+10  
4   -6.046559e+08 -9.020766e+08 -1.129994e+08  
..            ...           ...           ...  
151  7.718044e+09  4.765270e+09  5.240984e+09  
152 -3.657770e+09 -9.308742e+09 -9.308742e+09  
153 -1.021386e+10 -5.631520e+10 -5.631520e+10  
154 -6.104003e+09 -8.227767e+09 -8.227767e+09  
155 -8.705863e+06 -1.829185e+07 -1.828956e+07  

[156 rows x 9 columns]

In [34]:
# 총자산증가율, 자기자본증가율, 매출액증가율, 총자본영업이익율, 총자본순이익율, 매출액순이익율, 매출액영업이익율, 자기자본비율, 부채비율, 총자본회전율, 자기자본회전율, 당기순이익, 영업활동으로인한현금흐름
def make_재무제표_파생변수(df):
  # 기업코드 unique 가져오기
  unique_기업코드 = list(df['기업'].unique())

  # 총자산증가율=[]
  # 자기자본증가율=[]
  # 매출액증가율=[]
  # 총자본영업이익율=[]
  # 총자본순이익율=[]
  # 매출액순이익율=[]
  # 매출액영업이익율=[]
  # 자기자본비율=[]
  # 부채비율=[]
  # 총자본회전율=[]
  # 자기자본회전율=[]
  # 당기순이익=[]
  # 영업활동으로인한현금흐름=[]

  tmp_total = pd.DataFrame()

  for i in unique_기업코드:
    
    tmp = df[df['기업']==i]
    
    # 총자산증가율 - 수정완료
    tmp['총자산증가율'] = tmp['자산총계'].diff(1) / tmp['자산총계'].shift(1) * 100

    # 자기자본증가율 - 수정완료
    tmp['자기자본증가율'] = (tmp['자본총계'] - tmp['자본총계'].shift(1))/tmp['자본총계'] * 100

    # 매출액증가율 - 수정완료
    tmp['매출액증가율'] = (tmp['매출액'] - tmp['매출액'].shift(1)) / tmp['매출액'].shift(1) * 100 

    # 총자본영업이익율 - 얼추비슷한 비율로 나온다. -> 다르다
    tmp['총자본영업이익율'] = (tmp['영업이익'] / tmp['자산총계']) * 100

    # 총자본순이익율 - 얼추비슷한 비율로 나온다. -> 다르다
    tmp['총자본순이익율'] = (tmp['당기순이익'] / tmp['자산총계']) * 100

    # 매출액순이익율 - 수정완료
    tmp['매출액순이익율'] = (tmp['당기순이익']/tmp['매출액'])*100

    # 매출액영업이익율 - 수정완료
    tmp['매출액영업이익율'] = (tmp['영업이익'] / tmp['매출액']) * 100

    # 자기자본비율 - 수정완료
    tmp['자기자본비율'] = (tmp['자본총계'] / tmp['자산총계']) * 100

    # 부채비율 - 수정완료
    tmp['부채비율'] = (tmp['부채총계'] / tmp['자본총계']) * 100

    # 총자본회전율 - 이건값이 안맞게 나옴
    tmp['총자본회전율'] = (tmp['매출액'] / tmp['자산총계']) * 100

    # 자기자본회전율 - 이것도 값이 안맞게 나온다.
    tmp['자기자본회전율'] = (tmp['매출액'] / tmp['자본총계']) * 100

    # 당기순이익(손실)
    tmp['당기순이익(손실)'] = (tmp['당기순이익'])

    # array형태인데 shape 한번봐야겠다. 괄호 안에 있는지 없는지
    # 총자산증가율.append(tmp['총자산증가율'].values)
    # 자기자본증가율.append(tmp['자기자본증가율'].values)
    # 매출액증가율.append(tmp['매출액증가율'].values)
    # 총자본영업이익율.append(tmp['총자본영업이익율'].values)
    # 총자본순이익율.append(tmp['총자본순이익율'].values)
    # 매출액순이익율.append(tmp['매출액순이익율'].values)
    # 매출액영업이익율.append(tmp['매출액영업이익율'].values)
    # 자기자본비율.append(tmp['자기자본비율'].values)
    # 부채비율.append(tmp['부채비율'].values)
    # 총자본회전율.append(tmp['총자본회전율'].values)
    # 자기자본회전율.append(tmp['자기자본회전율'].values)
    

    tmp = tmp[['기업', '연도', '총자산증가율', '자기자본증가율','매출액증가율','총자본영업이익율','총자본순이익율','매출액순이익율','매출액영업이익율', '자기자본비율', '부채비율', '총자본회전율', '자기자본회전율', '당기순이익(손실)']]
    tmp_total = pd.concat([tmp_total, tmp], axis=0)
  # df['자산총계'].diff(1)

  return tmp_total



tmp_total = make_재무제표_파생변수(null_fill_data)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:33: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stab

In [35]:
tmp_total.head(30)

기업    연도      총자산증가율     자기자본증가율      매출액증가율   총자본영업이익율     총자본순이익율  \
0   006730  2017         NaN         NaN         NaN  -0.932333   -1.767489   
1   006730  2018   14.780118   25.422116   87.868600  -0.000894   -0.601708   
2   006730  2019    5.018489    3.565213   18.345080   0.797204    0.002600   
3   006730  2020    0.384633   -3.820188  -34.260713  -0.356007   -0.925143   
4   018700  2017         NaN         NaN         NaN  -1.493897   -2.228721   
5   018700  2018   -4.879579   -5.590801  149.498529  -5.245811  -14.654924   
6   018700  2019   -3.492879   -5.101591  -59.848235  -8.936651  -19.045420   
7   018700  2020   37.768565   16.449343  602.879489   1.780440  -16.693827   
8   019570  2018         NaN         NaN         NaN  -4.351808   -6.230429   
9   019570  2019   52.392984   27.867087         NaN -12.870363  -12.568642   
10  019570  2020  125.392705    8.554741         NaN -18.783385  -19.006111   
11  020400  2017         NaN         NaN         NaN   2.316312    1.206283   
12  020400  2018   22.603525   16.574270    8.236393   2.240091    1.011523   
13  020400  2019   25.230692    4.371414    9.480848   3.062305    0.000715   
14  020400  2020   -0.306808    3.798486   25.517953   2.769316    1.434617   
15  021080  2017         NaN         NaN         NaN  15.641595   11.088253   
16  021080  2018   -0.131111    5.840953  -31.878147  10.211726    8.935881   
17  021080  2019    7.088884    5.455256   25.263981  10.140784    8.021565   
18  021080  2020   19.358317   13.297236   50.126725  17.129962   14.564519   
19  023460  2017         NaN         NaN         NaN   0.775356    0.989685   
20  023460  2018    7.367341    0.010516   34.773830   1.492355    0.964211   
21  023460  2019    4.030600    6.556407  151.504148   7.664552    5.822476   
22  023460  2020   13.099331   13.656992   83.886420  18.929634   15.531005   
23  035290  2017         NaN         NaN         NaN -21.713222  -47.441741   
24  035290  2018  149.915627    5.497047  -32.194753 -10.734273   -9.525771   
25  035290  2019  -62.102253  377.279688  -11.954661 -44.371527 -154.590903   
26  035290  2020   28.193055  134.517314  -45.847115 -11.355491    7.927286   
27  036170  2017         NaN         NaN         NaN   2.957649    0.600024   
28  036170  2018   16.734854    8.671842   -8.860762   1.149173   -2.262028   
29  036170  2019  -10.836083   11.048664   -4.942337  -0.000956  -15.070402   

       매출액순이익율   매출액영업이익율     자기자본비율        부채비율      총자본회전율     자기자본회전율  \
0   -46.153355 -24.345440  32.183772  210.715598    3.829601   11.899169   
1    -9.599407  -0.014264  37.597608  165.974364    6.268175   16.671738   
2     0.036810  11.286099  37.124510  169.363825    7.063591   19.026759   
3   -19.999794  -7.696186  35.621458  180.729669    4.625762   12.985887   
4   -25.774059 -17.276179  93.557112    6.886582    8.647148    9.242641   
5   -64.612484 -23.128394  93.148735    7.355189   22.681258   24.349507   
6  -201.826478 -94.702701  91.835017    8.890926    9.436532   10.275527   
7   -34.674703   3.698148  79.782622   25.340579   48.144109   60.344105   
8          NaN        NaN  85.395283   17.102487         NaN         NaN   
9          NaN        NaN  77.684690   28.725492         NaN         NaN   
10         NaN        NaN  37.690716  165.317327         NaN         NaN   
11    1.120333   2.151270  39.792626  151.302842  107.671827  270.582362   
12    1.064152   2.356640  38.904481  157.039799   95.054446  244.327758   
13    0.000861   3.685094  32.486365  207.821454   83.099768  255.798913   
14    1.371184   2.646869  33.873003  195.220353  104.626129  308.877627   
15   34.708999  48.962097  90.208248   10.854608   31.946334   35.413984   
16   41.007199  46.862115  95.929896    4.242789   21.791006   22.715552   
17   31.470237  39.784368  94.748468    5.542604   25.489369   26.902144   
18   45.428850  53.430839  91.555949    9.222832   32.060066   35.016912   
19   20.780257  16.280018  74.623698  

In [36]:
tmp_total[tmp_total['기업']=='284620']

기업    연도      총자산증가율    자기자본증가율  매출액증가율   총자본영업이익율    총자본순이익율  \
102  284620  2018         NaN        NaN     NaN  -0.244517   0.367440   
103  284620  2019    0.923711   0.808295     NaN  -0.412974   0.748611   
104  284620  2020  188.394278  19.794756     NaN -27.284750 -43.242810   

     매출액순이익율  매출액영업이익율     자기자본비율        부채비율  총자본회전율  자기자본회전율     당기순이익(손실)  
102      inf      -inf  92.716036    7.856208     0.0      0.0  3.894665e+07  
103      inf      -inf  92.616057    7.972639     0.0      0.0  8.008156e+07  
104     -inf      -inf  40.040260  149.748631     0.0      0.0 -1.334065e+10

# 기존 df와 merge작업
- left join 이후 np.where로 null값 채우기

In [37]:
# 2017년이 아닌 연도만 기준으로(2017년 데이터는 변화율 구하기 위해서 사용한것임.)
tmp_total = tmp_total[tmp_total['연도']!=2017].reset_index(drop=True)

# 일단 inf나온것은 nan으로 모두 대체한다.
tmp_total.replace(np.inf, np.nan, inplace = True)
tmp_total.replace(-np.inf, np.nan, inplace = True)

In [38]:
df

KIS   Stock           Name  총자산증가율 (2018)  자기자본증가율 (2018)  \
0     036644  389500      (주)에스비비테크          46.03          -39.46   
1     497730  296640        (주)이노룰스          81.84          215.02   
2     499459  417970       모델솔루션(주)          23.53           44.46   
3     HV2035  393890       더블유씨피(주)          69.90           68.28   
4     HS8995  314140       (주)알피바이오          38.75           31.69   
...      ...     ...            ...            ...             ...   
1396  I73639  950110  SBI핀테크솔루션즈(주)         146.31           77.94   
1397  125028  046970         (주)우리로          42.14            8.83   
1398  383046  113810         디젠스(주)         -12.94            6.17   
1399  F11411  155650        와이엠씨(주)         -22.34           13.57   
1400  099145  149950         (주)아바텍          -2.81           -0.35   

      매출액증가율 (2018)  총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  \
0              2.26            -5.23           -9.21         -168.76   
1             -6.23             3.88            1.43           68.58   
2              9.65            25.30           19.34          768.48   
3               NaN            -8.72           -7.59         -559.62   
4             -5.02             8.28            5.66          146.00   
...             ...              ...             ...             ...   
1396          10.83             8.05            7.49           24.84   
1397          92.45             4.43           -0.55           -2.82   
1398         -29.81            -3.98          -16.02          -87.40   
1399          59.47            10.09            7.46          196.38   
1400         -22.27             3.76            5.22           94.76   

      매출액순이익율 (2018)  매출액영업이익율 (2018)  자기자본비율 (2018)  부채비율 (2018)  \
0             -20.80           -11.80          20.22       394.44   
1               1.90             5.16          40.95       144.17   
2              12.97            16.97          47.94       108.60   
3                NaN              NaN          93.30         7.18   
4               6.74             9.85          58.82        70.00   
...              ...              ...            ...          ...   
1396           16.69            17.92          45.68       118.91   
1397           -1.06             8.50          48.09       107.95   
1398          -31.83            -7.91          38.61       158.99   
1399            6.33             8.55          50.08        99.67   
1400           10.39             7.49          93.86         6.55   

      유보액/납입자본비율 (2018)  총자본회전율 (2018)  자기자본회전율 (2018)  당기순이익(손실) (2018)  \
0                340.07           0.44            1.39     -1.917031e+09   
1               2371.70           0.75            2.16      1.770280e+08   
2               2005.45           1.49            3.33      7.210181e+09   
3               8560.18            NaN             NaN     -9.513513e+09   
4               1558.22           0.84            1.40      4.363446e+09   
...                 ...            ...             ...               ...   
1396              93.99           0.45            0.88      3.963095e+09   
1397             194.50           0.52            0.96     -3.391840e+08   
1398              63.05           0.50            1.44     -1.185408e+10   
1399             800.95           1.18            2.87      1.464977e+10   
1400            1650.60           0.50            0.54      7.726745e+09   

      영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  자기자본증가율 (2019)  매출액증가율 (2019)  \
0            6.075800e+07         -21.16          -95.80         -34.70   
1            1.148444e+09          -1.20           32.63          43.95   
2            5.469434e+09           9.91           24.51           8.93   
3           -8.218473e+09         172.97           31.73            NaN   
4            6.995499e+09          29.38            7.64           7.86   
...                   ...            ...             ...            ...   
1396         2.682301e+09           

In [39]:
tmp_total.head(30)

기업    연도      총자산증가율     자기자본증가율      매출액증가율   총자본영업이익율     총자본순이익율  \
0   006730  2018   14.780118   25.422116   87.868600  -0.000894   -0.601708   
1   006730  2019    5.018489    3.565213   18.345080   0.797204    0.002600   
2   006730  2020    0.384633   -3.820188  -34.260713  -0.356007   -0.925143   
3   018700  2018   -4.879579   -5.590801  149.498529  -5.245811  -14.654924   
4   018700  2019   -3.492879   -5.101591  -59.848235  -8.936651  -19.045420   
5   018700  2020   37.768565   16.449343  602.879489   1.780440  -16.693827   
6   019570  2018         NaN         NaN         NaN  -4.351808   -6.230429   
7   019570  2019   52.392984   27.867087         NaN -12.870363  -12.568642   
8   019570  2020  125.392705    8.554741         NaN -18.783385  -19.006111   
9   020400  2018   22.603525   16.574270    8.236393   2.240091    1.011523   
10  020400  2019   25.230692    4.371414    9.480848   3.062305    0.000715   
11  020400  2020   -0.306808    3.798486   25.517953   2.769316    1.434617   
12  021080  2018   -0.131111    5.840953  -31.878147  10.211726    8.935881   
13  021080  2019    7.088884    5.455256   25.263981  10.140784    8.021565   
14  021080  2020   19.358317   13.297236   50.126725  17.129962   14.564519   
15  023460  2018    7.367341    0.010516   34.773830   1.492355    0.964211   
16  023460  2019    4.030600    6.556407  151.504148   7.664552    5.822476   
17  023460  2020   13.099331   13.656992   83.886420  18.929634   15.531005   
18  035290  2018  149.915627    5.497047  -32.194753 -10.734273   -9.525771   
19  035290  2019  -62.102253  377.279688  -11.954661 -44.371527 -154.590903   
20  035290  2020   28.193055  134.517314  -45.847115 -11.355491    7.927286   
21  036170  2018   16.734854    8.671842   -8.860762   1.149173   -2.262028   
22  036170  2019  -10.836083   11.048664   -4.942337  -0.000956  -15.070402   
23  036170  2020    3.154832   -4.676191   -0.358003  -0.475048  -11.072020   
24  037460  2018   33.050766   10.117261   41.147167  -2.124703    1.822527   
25  037460  2019    3.538793    0.002944   19.879994  -2.213968   -0.339231   
26  037460  2020   -2.627397    8.711853   39.671217   2.491919    4.637085   
27  042370  2018   -4.051807    0.491882  -87.244051   0.755279    0.369731   
28  042370  2019   -0.963460    1.663734  178.267651   1.956893    0.762507   
29  042370  2020    5.691760    8.058586   22.821541   3.557399    2.900238   

       매출액순이익율   매출액영업이익율     자기자본비율        부채비율      총자본회전율     자기자본회전율  \
0    -9.599407  -0.014264  37.597608  165.974364    6.268175   16.671738   
1     0.036810  11.286099  37.124510  169.363825    7.063591   19.026759   
2   -19.999794  -7.696186  35.621458  180.729669    4.625762   12.985887   
3   -64.612484 -23.128394  93.148735    7.355189   22.681258   24.349507   
4  -201.826478 -94.702701  91.835017    8.890926    9.436532   10.275527   
5   -34.674703   3.698148  79.782622   25.340579   48.144109   60.344105   
6          NaN        NaN  85.395283   17.102487         NaN         NaN   
7          NaN        NaN  77.684690   28.725492         NaN         NaN   
8          NaN        NaN  37.690716  165.317327         NaN         NaN   
9     1.064152   2.356640  38.904481  157.039799   95.054446  244.327758   
10    0.000861   3.685094  32.486365  207.821454   83.099768  255.798913   
11    1.371184   2.646869  33.873003  195.220353  104.626129  308.877627   
12   41.007199  46.862115  95.929896    4.242789   21.791006   22.715552   
13   31.470237  39.784368  94.748468    5.542604   25.489369   26.902144   
14   45.428850  53.430839  91.555949    9.222832   32.060066   35.016912   
15   16.128442  24.962755  69.510472   43.863215    5.978325    8.600611   
16   40.285068  53.030188  71.505526   39.849332   14.453186   20.212684   
17   66.091614  80.554352  73.223822   36.567578   23.499207   32.092298   
18  -46.608699 -52.521786  32.649292  206.285354   20.437754   62.597847   
19 -325.580058 -93.449771 -31.070076 -

In [41]:
df.to_csv('df_원본.csv', index=False)
tmp_total.to_csv('opendart.csv', index = False)

In [ ]:
tmp_total.describe()

총자산증가율      자기자본증가율         매출액증가율    총자본영업이익율     총자본순이익율  \
count   88.000000    88.000000      68.000000  135.000000  135.000000   
mean    42.507364    29.026544    2235.465013   -9.988864  -20.983598   
std    112.720782   177.253779   18163.158579   22.763095   59.103207   
min    -82.551375  -126.747407    -100.000000 -118.470090 -622.358107   
25%     -4.240969    -5.223894     -33.161686  -16.802564  -22.953491   
50%      3.784697     3.963705       1.416480   -4.462285   -9.525771   
75%     38.066292    21.201596      35.998177    1.479374    0.309771   
max    714.365493  1581.318537  149805.530667   58.943951   44.786053   

             매출액순이익율       매출액영업이익율      자기자본비율          부채비율      총자본회전율  \
count     107.000000     107.000000  134.000000    134.000000  127.000000   
mean    -5667.677294   -4907.532830   58.127098   -167.832422   32.639211   
std     35439.949420   34309.035488   88.286598   1626.502299   55.956746   
min   -353205.536152 -350051.850380 -690.981106 -16075.124467    0.000000   
25%      -363.085580    -261.718893   46.867990      6.583570    0.442905   
50%       -34.674703     -11.225950   76.605119     20.027772   13.113658   
75%        -0.045691       6.303471   92.566304     81.943026   41.693085   
max       309.991346      80.554352   99.635646    607.021922  477.772352   

            자기자본회전율     당기순이익(손실)  
count    126.000000  1.350000e+02  
mean    -244.681986 -8.173414e+09  
std     2233.421860  1.770613e+10  
min   -21213.969911 -1.172341e+11  
25%        0.236047 -1.120089e+10  
50%       13.315242 -4.045846e+09  
75%       49.508001  4.302014e+07  
max      860.731007  2.663508e+10

In [ ]:
df.describe()

총자산증가율 (2018)  자기자본증가율 (2018)  매출액증가율 (2018)  총자본영업이익율 (2018)  \
count   1.385000e+03     1358.000000    1367.000000      1400.000000   
mean    8.307525e+02      685.906127      21.573921         2.502879   
std     2.926893e+04    23315.797840     153.383154        15.853033   
min    -6.804000e+01      -90.030000    -100.000000      -129.530000   
25%    -1.530000e+00       -1.100000      -8.145000        -0.775000   
50%     7.340000e+00        6.950000       4.380000         2.905000   
75%     2.496000e+01       23.480000      20.830000         8.742500   
max     1.089023e+06   858889.270000    3648.630000       137.560000   

       총자본순이익율 (2018)  자본금순이익율 (2018)  매출액순이익율 (2018)  매출액영업이익율 (2018)  \
count     1401.000000     1401.000000     1384.000000      1384.000000   
mean        -1.536881       70.205332     -616.313439      -471.265383   
std         26.836112     2323.153378    11080.499725     10273.463568   
min       -532.050000   -64738.350000  -353205.520000   -350051.850000   
25%         -3.400000      -31.970000       -4.920000        -0.890000   
50%          2.620000       33.020000        3.670000         4.455000   
75%          7.760000      143.810000       10.500000        11.710000   
max        114.400000    52574.620000     1307.340000        77.620000   

       자기자본비율 (2018)  부채비율 (2018)  유보액/납입자본비율 (2018)  총자본회전율 (2018)  \
count    1382.000000  1382.000000        1401.000000    1375.000000   
mean       64.496852    97.685753        1697.744404       0.802269   
std        20.955324   299.279136        9641.966500       0.633225   
min         1.120000     0.910000       -8327.170000       0.010000   
25%        49.792500    21.345000         334.930000       0.400000   
50%        65.125000    53.550000         783.650000       0.660000   
75%        82.407500   100.827500        1592.300000       1.050000   
max        99.100000  8799.540000      248880.480000       8.440000   

       자기자본회전율 (2018)  당기순이익(손실) (2018)  영업활동으로인한현금흐름 (2018)  총자산증가율 (2019)  \
count     1359.000000      1.378000e+03         1.368000e+03    1399.000000   
mean         1.610412      2.820362e+09         5.502024e+09      21.805661   
std          2.491764      1.698619e+10         1.958206e+10      77.475601   
min          0.010000     -1.234088e+11        -1.672801e+11     -82.480000   
25%          0.580000     -2.380625e+09        -1.445045e+09      -1.220000   
50%          1.050000      1.959088e+09         2.313819e+09       7.260000   
75%          1.880000      6.327230e+09         8.111196e+09      22.825000   
max         60.310000      1.546099e+11         2.871206e+11    1628.770000   

       자기자본증가율 (2019)  매출액증가율 (2019)  총자본영업이익율 (2019)  총자본순이익율 (2019)  \
count     1364.000000    1381.000000      1400.000000     1399.000000   
mean        32.059802     116.543526         1.947186       -2.671001   
std        173.069248    3535.374771        15.856586       25.873139   
min        -98.300000    -100.000000      -191.540000     -353.270000   
25%         -1.172500      -8.560000        -0.417500       -3.975000   
50%          5.500000       5.300000         2.810000        2.210000   
75%         19.890000      21.100000         8.277500        7.240000   
max       3408.850000  131338.880000        89.070000       71.040000   

       자본금순이익율 (2019)  매출액순이익율 (2019)  매출액영업이익율 (2019)  자기자본비율 (2019)  \
count     1401.000000     1388.000000      1389.000000    1380.000000   
mean        79.691863     -372.374265      -111.402441      64.173964   
std       2077.977273     6496.959176      1348.932879      20.868991   
min     -25660.800000  -184682.910000    -32961.540000       1.080000   
25%        -35.650000       -7.087500        -0.550000      49.500000   
50%         28.120000        3.070000         4.470000      65.405000   
75%        125.840000       10.292500        11.550000      81.962500   
max      68510.460000     1085.710000        99.850000      99.570000   

       부채비율 (2019)  유보액

In [ ]:
# 채우기 이전에 df의 널값 총 개수
print(df.isna().sum().sum())

734


In [ ]:
# # 널값인 데이터만 보기
# df2 = df.copy()
# df2['NaN_cnt'] = df2.isnull().sum(1)

# # 재무제표 crawling이 필요한 코드
# stock_code = list(df2[df2['NaN_cnt']>0]['Stock'].values)

# 연도별로 나누기
df_base = df[['KIS', 'Stock', 'Name']]
df_2018 = pd.concat([df_base,df[[col for col in df.columns if '2018' in col]]], axis=1)
df_2019 = pd.concat([df_base,df[[col for col in df.columns if '2019' in col]]], axis=1)
df_2020 = pd.concat([df_base,df[[col for col in df.columns if '2020' in col]]], axis=1)
  
# 채우기 위해 크롤링한 open Dart df
temp_df_2018 = tmp_total[tmp_total['연도'] == 2018].reset_index(drop=True)
temp_df_2019 = tmp_total[tmp_total['연도'] == 2019].reset_index(drop=True)
temp_df_2020 = tmp_total[tmp_total['연도'] == 2020].reset_index(drop=True)


# 채우기 이전에 null값 개수
print(f'2018년도 채우기 이전 널값 개수: {df_2018.isna().sum().sum()}')
print(f'2019년도 채우기 이전 널값 개수: {df_2019.isna().sum().sum()}')
print(f'2020년도 채우기 이전 널값 개수: {df_2020.isna().sum().sum()}')

# df_2018.replace({np.NAN: 'none'}, inplace=True)

# # 채우기
# for i in range(len(df_2018)):
#   for j in df_2018.columns[3:]:
#     if df_2018.loc[i, j]== 'none':
#       stock_ccode = df_2018['Stock'][i]
#       try:
#         df_2018.loc[i,j]=
#       except:
#         pass

# df_2018.replace({'none', np.NAN}, inplace=True)

2018년도 채우기 이전 널값 개수: 290
2019년도 채우기 이전 널값 개수: 242
2020년도 채우기 이전 널값 개수: 202


In [ ]:
df_2018.head(30)

KIS   Stock          Name  총자산증가율 (2018)  자기자본증가율 (2018)  \
0   036644  389500     (주)에스비비테크          46.03          -39.46   
1   497730  296640       (주)이노룰스          81.84          215.02   
2   499459  417970      모델솔루션(주)          23.53           44.46   
3   HV2035  393890      더블유씨피(주)          69.90           68.28   
4   HS8995  314140      (주)알피바이오          38.75           31.69   
5   389238  368970       (주)오에스피            NaN             NaN   
6   IW9267  394280  오픈엣지테크놀로지(주)       24565.93        24767.99   
7   484917  148930    (주)에이치와이티씨           6.46             NaN   
8   894281  107600        (주)새빗켐          24.53           15.87   
9   867055  129920      (주)대성하이텍           8.36           26.30   
10  E66612  368600     (주)아이씨에이치          35.25          165.42   
11  IO9843  365340      성일하이텍(주)          42.94           47.71   
12  I91043  397030    (주)에이프릴바이오         137.78          -40.83   
13  J89214  328130         (주)루닛         154.08             NaN   
14  IO4904  403870    (주)에이치피에스피         -51.38          -43.67   
15  370082  112290      영창케미칼(주)           5.16            3.00   
16  166646  402030    (주)코난테크놀로지          18.78            2.69   
17  FZ6224  412350       레이저쎌(주)          25.26           42.92   
18  CR3994  310210       보로노이(주)         243.30          213.07   
19  CC9475  377460     (주)위니아에이드          63.52          820.48   
20  IW0365  362320      (주)청담글로벌         359.54          341.83   
21  J50025  399720       (주)가온칩스          65.42           73.30   
22  I66973  389260      대명에너지(주)         -30.60           12.01   
23  IT6914  389140       (주)포바이포         226.18           96.88   
24  G94198  388050       지투파워(주)          34.93           40.13   
25  219717  396300     (주)세아메카닉스          26.53           14.90   
26  F26242  366030       (주)공구우먼           8.65           26.41   
27  E82895  388720     (주)유일로보틱스          19.25            4.34   
28  L65171  288980      (주)모아데이타          40.62           83.50   
29  HL2329  376930         노을(주)          31.39           23.08   

    매출액증가율 (2018)  총자본영업이익율 (2018)  총자본순이익율 (2018)  자본금순이익율 (2018)  \
0            2.26            -5.23           -9.21         -168.76   
1           -6.23             3.88            1.43           68.58   
2            9.65            25.30           19.34          768.48   
3             NaN            -8.72           -7.59         -559.62   
4           -5.02             8.28            5.66          146.00   
5             NaN            21.66           15.28          284.44   
6             NaN            26.41           25.40         2683.41   
7           48.83            21.61           14.00           74.47   
8            5.81            14.03           10.14           51.67   
9            7.67             3.96            1.57          162.88   
10          -5.52            20.84           19.83          864.26   
11          50.40            19.89           16.20          318.06   
12            NaN           -11.24           -9.94         -109.36   
13          71.88           -83.72         -102.02        -1967.71   
14         -91.62           -14.48          -12.67         -126.71   
15          40.27             7.28            1.55           92.82   
16          16.39             4.00            1.58            8.01   
17           2.47           -68.09          -67.44         -145.40   
18        -100.00           -39.23          -40.08         -173.61   
19         124.61            11.58            8.63          439.49   
20        3173.33            58.38           57.31          311.91   
21          72.07            17.28           18.32         1002.90   
22          -4.95             7.56            5.38           63.43   
23         506.28            43.40           41.30          739.42   
24          89.58            19.33           16.79          512.59   
25          44.22             9.27           10.63          684.60   
26           2.53            24.59     

In [ ]:
# 널값인 데이터만 보기
df22 = df_2018.copy()
df22['NaN_cnt'] = df22.isnull().sum(1)

# 재무제표 crawling이 필요한 코드
stock_code_2018 = list(df22[df22['NaN_cnt']>0]['Stock'].values)
stock_code_2018


tmp_2018 = []
for i in stock_code_2018:
  if i in list(temp_df_2018['기업'].unique()):
    tmp_2018.append(i)
print(tmp_2018)

['299660', '307180', '309930', '298380', '290660', '208340', '115180', '042420', '042370', '006730', '023460', '021080', '019570', '270520', '277070', '291230', '284620']


In [44]:
df22 = df.copy()
df22['NaN_cnt'] = df22.isnull().sum(1)

# 재무제표 crawling이 필요한 코드
stock_code_null = list(df22[df22['NaN_cnt']>0]['Stock'].values)
stock_code_null

['393890',
 '368970',
 '394280',
 '148930',
 '107600',
 '397030',
 '328130',
 '310210',
 '362320',
 '389140',
 '388050',
 '376930',
 '330730',
 '276040',
 '298870',
 '377480',
 '357880',
 '376300',
 '389030',
 '261780',
 '308080',
 '315640',
 '376980',
 '195940',
 '377030',
 '365270',
 '352910',
 '187660',
 '363250',
 '347700',
 '338220',
 '334970',
 '340930',
 '321820',
 '236810',
 '335810',
 '354200',
 '317690',
 '348150',
 '304840',
 '331920',
 '950200',
 '330350',
 '339950',
 '288330',
 '235980',
 '335890',
 '190650',
 '216080',
 '317240',
 '100790',
 '299660',
 '314130',
 '307180',
 '310870',
 '309930',
 '298380',
 '307750',
 '307870',
 '291650',
 '294090',
 '323990',
 '298060',
 '290660',
 '027360',
 '208340',
 '306620',
 '244460',
 '300080',
 '293780',
 '322780',
 '307930',
 '319400',
 '227950',
 '217730',
 '115180',
 '228760',
 '239890',
 '241820',
 '241520',
 '246690',
 '267320',
 '027830',
 '054220',
 '058450',
 '007330',
 '016600',
 '036170',
 '037460',
 '042420',
 '042370',

In [46]:
print(len(stock_code_null))

117


In [47]:
len(df)

1401

In [ ]:
# 기업 연도 한개씩으로 내가 새로 만들어보기
# 정민이한테 물어보기
stock_code = '284620'
year = 2018
indicators = ['자산총계', '부채총계', '자본총계', '매출액', '영업이익', '당기순이익', '법인세차감전 순이익']


result = dart.finstate(stock_code, year,'11011') # 사업보고서 데이터 가져오기
# if report is None: # 리포트가 없다면 (참고 : 리포트가 없으면 None을 반환함)
result = result.loc[result['fs_nm'] =='재무제표'] # 연결재무제표 아닌 재무제표 필터링
result = result.loc[result['account_nm'].isin(indicators)]
data = [] 
for y, c in zip([year], ['thstrm_amount']):
  record = [stock_code, y]
  for indic in indicators:
    # account_nm이 indic인 행의 c 컬럼 값을 가져옴
    if sum(result['account_nm']==indic)>0: # 해당 지표가 있다면 추가(지표가 없는 경우가 있음)
      value = result.loc[result['account_nm']==indic, c].iloc[0]
    else:
      value = np.nan

    record.append(value)

  data.append(record)
tt= pd.DataFrame(data, columns = ['기업', '연도'] + indicators)  
tt
result

rcept_no reprt_code bsns_year corp_code stock_code fs_div fs_nm  \
2   20190401004592      11011      2018  01263527     284620    OFS  재무제표   
5   20190401004592      11011      2018  01263527     284620    OFS  재무제표   
8   20190401004592      11011      2018  01263527     284620    OFS  재무제표   
9   20190401004592      11011      2018  01263527     284620    OFS  재무제표   
10  20190401004592      11011      2018  01263527     284620    OFS  재무제표   
11  20190401004592      11011      2018  01263527     284620    OFS  재무제표   
12  20190401004592      11011      2018  01263527     284620    OFS  재무제표   

   sj_div  sj_nm  account_nm thstrm_nm                thstrm_dt  \
2      BS  재무상태표        자산총계     제 2 기            2018.12.31 현재   
5      BS  재무상태표        부채총계     제 2 기            2018.12.31 현재   
8      BS  재무상태표        자본총계     제 2 기            2018.12.31 현재   
9      IS  손익계산서         매출액     제 2 기  2018.01.01 ~ 2018.12.31   
10     IS  손익계산서        영업이익     제 2 기  2018.01.01 ~ 2018.12.31   
11     IS  손익계산서  법인세차감전 순이익     제 2 기  2018.01.01 ~ 2018.12.31   
12     IS  손익계산서       당기순이익     제 2 기  2018.01.01 ~ 2018.12.31   

     thstrm_amount frmtrm_nm                frmtrm_dt  frmtrm_amount  \
2   10,599,449,914     제 1 기            2017.12.31 현재  1,690,775,632   
5      772,060,139     제 1 기            2017.12.31 현재    740,190,004   
8    9,827,389,775     제 1 기            2017.12.31 현재    950,585,628   
9                0     제 1 기  2017.11.03 ~ 2017.12.31              0   
10     -25,917,490     제 1 기  2017.11.03 ~ 2017.12.31     -7,044,640   
11      49,931,598     제 1 기  2017.11.03 ~ 2017.12.31     -9,495,262   
12      38,946,647     제 1 기  2017.11.03 ~ 2017.12.31     -7,406,305   

   bfefrmtrm_amount ord currency  
2                 -   6      KRW  
5                 -  12      KRW  
8                 -  22      KRW  
9                 -  24      KRW  
10                -  26      KRW  
11                -  28      KRW  
12                -  30      KRW

In [ ]:
temp_df_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   기업         31 non-null     object 
 1   연도         31 non-null     object 
 2   총자산증가율     20 non-null     float64
 3   자기자본증가율    20 non-null     float64
 4   매출액증가율     18 non-null     float64
 5   총자본영업이익율   31 non-null     float64
 6   총자본순이익율    31 non-null     float64
 7   매출액순이익율    23 non-null     float64
 8   매출액영업이익율   23 non-null     float64
 9   자기자본비율     31 non-null     float64
 10  부채비율       31 non-null     float64
 11  총자본회전율     29 non-null     float64
 12  자기자본회전율    29 non-null     float64
 13  당기순이익(손실)  31 non-null     float64
dtypes: float64(12), object(2)
memory usage: 3.5+ KB


In [ ]:
df_2018.loc[0,'총자산증가율 (2018)']

46.03

In [ ]:
temp_df_2018.columns

Index(['기업', '연도', '총자산증가율', '자기자본증가율', '매출액증가율', '총자본영업이익율', '총자본순이익율',
       '매출액순이익율', '매출액영업이익율', '자기자본비율', '부채비율', '총자본회전율', '자기자본회전율',
       '당기순이익(손실)'],
      dtype='object')

In [ ]:
df_2018.columns

Index(['KIS', 'Stock', 'Name', '총자산증가율 (2018)', '자기자본증가율 (2018)',
       '매출액증가율 (2018)', '총자본영업이익율 (2018)', '총자본순이익율 (2018)', '자본금순이익율 (2018)',
       '매출액순이익율 (2018)', '매출액영업이익율 (2018)', '자기자본비율 (2018)', '부채비율 (2018)',
       '유보액/납입자본비율 (2018)', '총자본회전율 (2018)', '자기자본회전율 (2018)',
       '당기순이익(손실) (2018)', '영업활동으로인한현금흐름 (2018)'],
      dtype='object')

In [ ]:
print(tmp_total.shape)
tmp_total.head(30)


(218, 13)


기업    연도      총자산증가율     자기자본증가율      매출액증가율   총자본영업이익율     총자본순이익율  \
0   006730  2018   -0.668837   -2.689208   87.868600  -0.000894   -0.601708   
1   006730  2019    5.018489    3.565213   18.345080   0.797204    0.002600   
2   006730  2020    9.260552   -6.685650  -33.750694  -1.145787   -1.779551   
3   018700  2018   -3.934614  -34.399102   42.034220  -9.445938  -27.292588   
4   018700  2018    6.588345   12.502938   47.506839  -6.846527  -13.681486   
5   018700  2019   31.469350   35.435777  -40.431197  -4.871750  -24.928192   
6   018700  2019  -28.638205  -77.016848   13.806998 -18.286298  -27.292588   
7   018700  2020   31.864254   33.953258  299.544877   0.101342  -17.343509   
8   019570  2018    2.916121   -7.778908         NaN  -4.351808   -6.230429   
9   019570  2019   52.392984   27.867087         NaN -12.870363  -12.568642   
10  019570  2020  125.392705    8.554741         NaN -18.783385  -19.006111   
11  020400  2018   22.603525   16.574270    8.236393   2.240091    1.011523   
12  020400  2019   25.230692    4.371414    9.480848   3.062305    0.000715   
13  020400  2020   -0.306808    3.798486   25.517953   2.769316    1.434617   
14  021080  2018   -0.131111    5.840953  -31.878147  10.211726    8.935881   
15  021080  2019    7.088884    5.455256   25.263981  10.140784    8.021565   
16  021080  2020   19.358317   13.297236   50.126725  17.129962   14.564519   
17  023460  2018   15.829438   -5.471420   12.484617   0.462781    0.195556   
18  023460  2018  -14.192328    0.000000    0.000000   0.539323    0.227900   
19  023460  2019   54.994858    9.983084  168.046286   4.347835    2.953773   
20  023460  2019   -6.746311    1.791516    0.000000   4.528593    3.412452   
21  023460  2020   19.870236   18.430078   83.997368   9.969460    5.278207   
22  035290  2018  140.302864   -6.120068  -32.194753 -10.677350  -13.529560   
23  035290  2019  -60.997376  347.004052  -11.954661 -44.371527 -154.591298   
24  035290  2020   28.193055  135.362647  -45.847115 -11.355491    7.927286   
25  036170  2018   17.132584    9.038674  -13.136714   0.643622   -2.097513   
26  036170  2019   -8.257494   14.194694   -6.413148  -0.009318  -13.471194   
27  036170  2020    2.914148   -4.799577    0.629797  -0.942799  -10.990388   
28  037460  2018   21.335795   11.173023   10.346566   5.715380    5.237654   
29  037460  2019   -1.486585    8.563865    3.845103   6.469389    6.067504   

       매출액순이익율    매출액영업이익율     자기자본비율        부채비율      총자본회전율     자기자본회전율  
0    -9.599407   -0.014264  37.597608  165.974364    6.268175   16.671738  
1     0.036810   11.286099  37.124510  169.363825    7.063591   19.026759  
2   -41.549607  -26.752264  31.848672  213.984832    4.282955   13.447828  
3  -163.494957  -56.585444  68.824122   45.297895   16.693229   24.254911  
4   -59.223010  -29.636542  73.796793   35.507244   23.101639   31.304395  
5  -238.151662  -46.542302  86.940282   15.021481   10.467360   12.039712  
6  -163.494957 -109.543203  68.824122   45.297895   16.693229   24.254911  
7   -34.289260    0.200359  79.024580   26.542906   50.580002   64.005405  
8          NaN         NaN  85.395283   17.102487         NaN         NaN  
9          NaN         NaN  77.684690   28.725492         NaN         NaN  
10         NaN         NaN  37.690716  165.317327         NaN         NaN  
11    1.064152    2.356640  38.904481  157.039799   95.054446  244.327758  
12    0.000861    3.685094  32.486365  207.821454   83.099768  255.798913  
13    1.371184    2.646869  33.873003  195.220353  104.626129  308.877627  
14   41.007199   46.862115  95.929896    4.242789   21.791006   22.715552  
15   31.470237   39.784368  94.748468    5.542604   25.489369   26.902144  
16   45.428850   53.430839  91.555949    9.222832   32.060066   35.016912  
17    1.171915    2.773324  25.870334  286.543139   16.686863   64.501924  
18    1.171915    2.773324  30.149208  231.683668   19.446819   64.501924  
19    8.782862   12.928019  21.608991 

# 최종 파일 만들기
- nan이 있는 종목 drop하기
- nan이 있는 데이터 117개 drop
- 1284개 기업 데이터로 확정

In [51]:
df22 = df.copy()
print(df22.isna().sum().sum())
del_list = list(df2[df2['NaN_cnt']>0].index)

734


In [52]:
df22.drop(del_list, axis=0, inplace=True)
df22.isna().sum()

KIS                    0
Stock                  0
Name                   0
총자산증가율 (2018)          0
자기자본증가율 (2018)         0
매출액증가율 (2018)          0
총자본영업이익율 (2018)        0
총자본순이익율 (2018)         0
자본금순이익율 (2018)         0
매출액순이익율 (2018)         0
매출액영업이익율 (2018)        0
자기자본비율 (2018)          0
부채비율 (2018)            0
유보액/납입자본비율 (2018)      0
총자본회전율 (2018)          0
자기자본회전율 (2018)         0
당기순이익(손실) (2018)       0
영업활동으로인한현금흐름 (2018)    0
총자산증가율 (2019)          0
자기자본증가율 (2019)         0
매출액증가율 (2019)          0
총자본영업이익율 (2019)        0
총자본순이익율 (2019)         0
자본금순이익율 (2019)         0
매출액순이익율 (2019)         0
매출액영업이익율 (2019)        0
자기자본비율 (2019)          0
부채비율 (2019)            0
유보액/납입자본비율 (2019)      0
총자본회전율 (2019)          0
자기자본회전율 (2019)         0
당기순이익(손실) (2019)       0
영업활동으로인한현금흐름 (2019)    0
총자산증가율 (2020)          0
자기자본증가율 (2020)         0
매출액증가율 (2020)          0
총자본영업이익율 (2020)        0
총자본순이익율 (2020)         0
자본금순이익율 (2020)         0
매출액순이익율 (2020)         0


In [53]:
len(df22)

1284

In [54]:
print(df22['2019_관리종목여부'].value_counts())
print(df22['2020_관리종목여부'].value_counts())
print(df22['2021_관리종목여부'].value_counts())

0    1245
1      39
Name: 2019_관리종목여부, dtype: int64
0    1252
1      32
Name: 2020_관리종목여부, dtype: int64
0    1259
1      25
Name: 2021_관리종목여부, dtype: int64


In [56]:
# 최종파일 내보내기
df22.reset_index(drop=True, inplace= True)
df22.to_csv('real_df.csv', index=False)

In [ ]:
# https://gils-lab.tistory.com/38?category=504501
